cms/src/builder/utils/kw.rs

```diff
@@ -0,0 +1,571 @@
+//! Key wrap module
+//!
+//! This module contains the key wrapping logic based on aes-kw algorithms
+//!
+
+use alloc::{string::String, vec::Vec};
+
+use crate::builder::{ContentEncryptionAlgorithm, Error, Result};
+use aes_kw::Kek;
+use const_oid::ObjectIdentifier;
+use der::Any;
+use spki::AlgorithmIdentifierOwned;
+
+/// Represents supported key wrap algorithm for ECC - as defined in [RFC 5753 Section 7.1.5].
+///
+/// As per [RFC 5753 Section 8]:
+/// ```text
+/// Implementations that support EnvelopedData with the ephemeral-static
+/// ECDH standard primitive:
+///
+/// - MUST support the dhSinglePass-stdDH-sha256kdf-scheme key
+///   agreement algorithm, the id-aes128-wrap key wrap algorithm, and
+///   the id-aes128-cbc content encryption algorithm; and
+/// - MAY support the dhSinglePass-stdDH-sha1kdf-scheme, dhSinglePass-
+///    stdDH-sha224kdf-scheme, dhSinglePass-stdDH-sha384kdf-scheme, and
+///    dhSinglePass-stdDH-sha512kdf-scheme key agreement algorithms;
+///    the id-alg-CMS3DESwrap, id-aes192-wrap, and id-aes256-wrap key
+///    wrap algorithms; and the des-ede3-cbc, id-aes192-cbc, and id-
+///    aes256-cbc content encryption algorithms; other algorithms MAY
+///    also be supported.
+/// ```
+///
+/// As such the following algorithm are currently supported
+/// - id-aes128-wrap
+/// - id-aes192-wrap
+/// - id-aes256-wrap
+///
+/// [RFC 5753 Section 8]: https://datatracker.ietf.org/doc/html/rfc5753#section-8
+/// [RFC 5753 Section 7.1.5]: https://datatracker.ietf.org/doc/html/rfc5753#section-7.1.5
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum KeyWrapAlgorithm {
+    /// id-aes128-wrap
+    Aes128,
+    /// id-aes192-wrap
+    Aes192,
+    /// id-aes256-wrap
+    Aes256,
+}
+impl KeyWrapAlgorithm {
+    /// Return the Object Identifier (OID) of the algorithm.
+    ///
+    /// OID are defined in [RFC 3565 Section 2.3.2]
+    ///
+    /// [RFC 3565 Section 2.3.2]:
+    /// ```text
+    /// NIST has assigned the following OIDs to define the AES key wrap
+    /// algorithm.
+    ///
+    ///     id-aes128-wrap OBJECT IDENTIFIER ::= { aes 5 }
+    ///     id-aes192-wrap OBJECT IDENTIFIER ::= { aes 25 }
+    ///     id-aes256-wrap OBJECT IDENTIFIER ::= { aes 45 }
+    ///
+    /// In all cases the parameters field MUST be absent.
+    /// ```
+    ///
+    /// [RFC 3565 Section 2.3.2]: https://datatracker.ietf.org/doc/html/rfc3565#section-2.3.2
+    fn oid(&self) -> ObjectIdentifier {
+        match self {
+            Self::Aes128 => const_oid::db::rfc5911::ID_AES_128_WRAP,
+            Self::Aes192 => const_oid::db::rfc5911::ID_AES_192_WRAP,
+            Self::Aes256 => const_oid::db::rfc5911::ID_AES_256_WRAP,
```

Comment:
TODO: We should implement AssociatedOid in https://github.com/RustCrypto/key-wraps/tree/master/aes-kw

Reply:
https://github.com/RustCrypto/key-wraps/pull/35

Reply:
Thanks will merge once its merged upstream

Reply:
we will need to move that to a generic parameter as well

Reply:
I could add a trait for Key Wrapping too. Do you want the trait to only implement AssociateOid, or the key wrapping logic itself (a keen to kdf for KeyAgreement trait)?

Reply:
> we will need to move that to a generic parameter as well

@baloo, I came up with a trait-based approach for key-wrapping, a first draft is available here https://github.com/nemynm/formats/pull/3, let me know your thoughts and if this is the kind of things you had in mind.